Address (some) warnings from golint and go vet

`golint` generates a lot of warnings, and `go vet` generates a few warnings too (see attached file)
[golint.txt](https://github.com/dgraph-io/badger/files/1245134/golint.txt).

We don’t need to clean up all the warnings, but it is worth cleaning up some of them.

Nice work. I just ran go lint again, and there're some comment related warnings that we should be able to tackle as well.

```
iterator.go:94:6: exported type IteratorOptions should have comment or be unexported
iterator.go:100:5: exported var DefaultIteratorOptions should have comment or be unexported
kv.go:141:5: exported var ErrInvalidDir should have comment or be unexported
kv.go:142:5: exported var ErrValueLogSize should have comment or be unexported
kv.go:143:5: exported var ErrExceedsMaxKeyValueSize should have comment or be unexported
kv.go:1039:5: exported var ErrNoRoom should have comment or be unexported
manifest.go:42:6: exported type Manifest should have comment or be unexported
manifest.go:60:6: exported type LevelManifest should have comment or be unexported
manifest.go:64:6: exported type TableManifest should have comment or be unexported
manifest.go:84:2: exported const ManifestFilename should have comment (or a comment on this block) or be unexported
manifest.go:107:1: exported function OpenOrCreateManifestFile should have comment or be unexported
manifest.go:107:48: exported func OpenOrCreateManifestFile returns unexported type *badger.manifestFile, which can be annoying to use
value.go:46:2: don't use underscores in Go names; const Bit_UNUSED should be BitUNUSED
value.go:51:5: error var Corrupt should have name of the form ErrFoo
value.go:51:5: exported var Corrupt should have comment or be unexported
value.go:52:5: error var CasMismatch should have name of the form ErrFoo
value.go:52:5: exported var CasMismatch should have comment or be unexported
value.go:53:5: error var KeyExists should have name of the form ErrFoo
value.go:53:5: exported var KeyExists should have comment or be unexported
```

Also, using `Err` prefix for all the error vars should be doable too (it should be OK to rename these public error variables, while we're still not tagged as v1.0).

--- a/iterator.go
+++ b/iterator.go
@@ -91,12 +91,14 @@ func (l *list) pop() *KVItem {
 	return i
 }
 
+// IteratorOptions is used to set options when iterating over Badger key-value stores.
 type IteratorOptions struct {
 	PrefetchSize int  // How many KV pairs to prefetch while iterating.
 	FetchValues  bool // Controls whether the values should be fetched from the value log.
 	Reverse      bool // Direction of iteration. False is forward, true is backward.
 }
 
+// DefaultIteratorOptions contains default options when iterating over Badger key-value stores.
 var DefaultIteratorOptions = IteratorOptions{
 	PrefetchSize: 100,
 	FetchValues:  true,

--- a/kv.go
+++ b/kv.go
@@ -138,8 +138,16 @@ type KV struct {
 	metricsTicker      *time.Ticker
 }
 
+// ErrInvalidDir is returned when Badger cannot find the directory
+// from where it is supposed to load the key-value store.
 var ErrInvalidDir = errors.New("Invalid Dir, directory does not exist")
+
+// ErrValueLogSize is returned when opt.ValueLogFileSize option is not within the valid
+// range.
 var ErrValueLogSize = errors.New("Invalid ValueLogFileSize, must be between 1MB and 1GB")
+
+// ErrExceedsMaxKeyValueSize is returned as part of Entry when the size of the key or value
+// exceeds the specified limits.
 var ErrExceedsMaxKeyValueSize = errors.New("Key (value) size exceeded 1MB (1GB) limit")
 
 const (
@@ -613,7 +621,7 @@ func (s *KV) writeToLSM(b *request) error {
 			}
 			// No need to decode existing value. Just need old CAS counter.
 			if oldValue.CASCounter != entry.CASCounterCheck {
-				entry.Error = CasMismatch
+				entry.Error = ErrCasMismatch
 				continue
 			}
 		}
@@ -626,7 +634,7 @@ func (s *KV) writeToLSM(b *request) error {
 			}
 			// Value already exists, don't write.
 			if exists {
-				entry.Error = KeyExists
+				entry.Error = ErrKeyExists
 				continue
 			}
 		}
@@ -912,7 +920,7 @@ func (s *KV) SetIfAbsent(key, val []byte, userMeta byte) error {
 	}
 	// Found the key, return KeyExists
 	if exists {
-		return KeyExists
+		return ErrKeyExists
 	}
 
 	e := &Entry{
@@ -1036,7 +1044,7 @@ func (s *KV) CompareAndDeleteAsync(key []byte, casCounter uint64, f func(error))
 	s.compareAsync(e, f)
 }
 
-var ErrNoRoom = errors.New("No room for write")
+var errNoRoom = errors.New("No room for write")
 
 // ensureRoomForWrite is always called serially.
 func (s *KV) ensureRoomForWrite() error {
@@ -1066,7 +1074,7 @@ func (s *KV) ensureRoomForWrite() error {
 		return nil
 	default:
 		// We need to do this to unlock and allow the flusher to modify imm.
-		return ErrNoRoom
+		return errNoRoom
 	}
 }
 

--- a/kv_test.go
+++ b/kv_test.go
@@ -717,7 +717,7 @@ func TestSetIfAbsent(t *testing.T) {
 	require.NoError(t, err)
 
 	err = kv.SetIfAbsent(key, []byte("val2"), 0x00)
-	require.EqualError(t, err, KeyExists.Error())
+	require.EqualError(t, err, ErrKeyExists.Error())
 }
 
 func BenchmarkExists(b *testing.B) {

--- a/manifest.go
+++ b/manifest.go
@@ -32,13 +32,14 @@ import (
 	"github.com/pkg/errors"
 )
 
+// Manifest represnts the contents of the MANIFEST file in a Badger store.
+//
 // The MANIFEST file describes the startup state of the db -- all LSM files and what level they're
 // at.
 //
 // It consists of a sequence of ManifestChangeSet objects.  Each of these is treated atomically,
 // and contains a sequence of ManifestChange's (file creations/deletions) which we use to
 // reconstruct the manifest at startup.
-
 type Manifest struct {
 	Levels []LevelManifest
 	Tables map[uint64]TableManifest
@@ -57,10 +58,14 @@ func createManifest() Manifest {
 	}
 }
 
+// LevelManifest contains information about LSM tree levels
+// in the MANIFEST file.
 type LevelManifest struct {
 	Tables map[uint64]struct{} // Set of table id's
 }
 
+// TableManifest contains information about a specific level
+// in the LSM tree.
 type TableManifest struct {
 	Level uint8
 }
@@ -81,6 +86,7 @@ type manifestFile struct {
 }
 
 const (
+	// ManifestFilename is the filename for the manifest file.
 	ManifestFilename                  = "MANIFEST"
 	manifestRewriteFilename           = "MANIFEST-REWRITE"
 	manifestDeletionsRewriteThreshold = 10000
@@ -104,6 +110,8 @@ func (m *Manifest) clone() Manifest {
 	return ret
 }
 
+// OpenOrCreateManifestFile opens a Badger manifest file if it exists, or creates on if
+// one doesn’t.
 func OpenOrCreateManifestFile(dir string) (ret *manifestFile, result Manifest, err error) {
 	return helpOpenOrCreateManifestFile(dir, manifestDeletionsRewriteThreshold)
 }

--- a/value.go
+++ b/value.go
@@ -43,14 +43,21 @@ import (
 const (
 	BitDelete       byte  = 1 // Set if the key has been deleted.
 	BitValuePointer byte  = 2 // Set if the value is NOT stored directly next to key.
-	Bit_UNUSED      byte  = 4
+	BitUnused       byte  = 4
 	BitSetIfAbsent  byte  = 8 // Set if the key is set using SetIfAbsent.
 	M               int64 = 1 << 20
 )
 
-var Corrupt = errors.New("Unable to find log. Potential data corruption")
-var CasMismatch = errors.New("CompareAndSet failed due to counter mismatch")
-var KeyExists = errors.New("SetIfAbsent failed since key already exists")
+// ErrCorrupt is returned when a value log file is corrupted.
+var ErrCorrupt = errors.New("Unable to find log. Potential data corruption")
+
+// ErrCasMismatch is returned when a CompareAndSet operation has failed due
+// to a counter mismatch.
+var ErrCasMismatch = errors.New("CompareAndSet failed due to counter mismatch")
+
+// ErrKeyExists is returned by SetIfAbsent metadata bit is set, but the
+// key already exists in the store.
+var ErrKeyExists = errors.New("SetIfAbsent failed since key already exists")
 
 const (
 	maxKeySize   = 1 << 20
@@ -711,7 +718,7 @@ func (vlog *valueLog) getFile(fid uint16) (*logFile, error) {
 		return vlog.files[idx].fid >= fid
 	})
 	if idx == len(vlog.files) || vlog.files[idx].fid != fid {
-		return nil, Corrupt
+		return nil, ErrCorrupt
 	}
 	return vlog.files[idx], nil
 }

